### PR TITLE
Dev met office agent within stack

### PR DESCRIPTION
**Pull request by markushofmeister**
Enhanced the agent slightly based on learnings from the Derived Information Agent and added several (integration) tests. This agent "design" shall be used as blueprint for reworking the other agents for the King's Lynn use case.

### CODE REVIEW

File: Agents/MetOfficeAgent/agent/kgutils/tsclient.py [hunk 1]
```diff
@@ -53,6 +54,19 @@ def __init__(self, kg_client, timeclass=TIMECLASS, rdb_url=DB_URL,
...
+    def connect(self):
+        """
+        Create context manager for RDB connection using getConnection method of Java
+        TimeSeries client (i.e. to ensure connection is closed after use)
+        """
+        try:
+            conn = self.connection.getConnection()
+            yield conn
+        finally:
+            conn.close()
```
[the remark this answers is not in the document]
Comment (markushofmeister) [reply Agents/MetOfficeAgent/agent/kgutils/tsclient.py]: Addressed by [71d6d8e](https://github.com/cambridge-cares/TheWorldAvatar/pull/407/commits/71d6d8ece6ca40ee001ae298de6f7a58f699d2b4)

File: Agents/MetOfficeAgent/README.md [hunk 2]
```diff
@@ -137,23 +151,64 @@ An overview of all provided API endpoints and their functionality is provided af
...
 [allows you to publish and install packages]: https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-apache-maven-registry#authenticating-to-github-packages
+[CMCL Docker registry wiki page]: https://github.com/cambridge-cares/TheWorldAvatar/wiki/Docker%3A-Image-registry
 [Common stack scripts]: https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Deploy/stacks/dynamic/common-scripts
 [Create SSH key]: https://docs.digitalocean.com/products/droplets/how-to/add-ssh-keys/create-with-openssh/
 [DataPoint]: https://www.metoffice.gov.uk/services/data/datapoint/about
-[Container registry on Github]: ghcr.io
+[Github container registry]: https://ghcr.io
 [Github package repository]: https://github.com/cambridge-cares/TheWorldAvatar/wiki/Packages
 [http://localhost:5000/]: http://localhost:5000/
 [Java Runtime Environment version >=11]: https://adoptopenjdk.net/?variant=openjdk8&jvmVariant=hotspot
```
Comment (gpeb2) [inline Agents/MetOfficeAgent/README.md:214]:
```suggestion
[Java Development Kit version >=11]: https://adoptopenjdk.net/?variant=openjdk8&jvmVariant=hotspot
```

File: Agents/MetOfficeAgent/README.md [hunk 3]
```diff
@@ -137,23 +151,64 @@ An overview of all provided API endpoints and their functionality is provided af
...
+    ```
+2. Install all required packages in virtual environment (the `-e` flag installs the project for-in place development):
+    ```bash
+    $ python -m pip install --upgrade pip
+    # Install all required packages from setup.py, incl. pytest etc.
+    python -m pip install -e .[dev]
+    # Install agentlogging (separate installation required, as not possible to include in setup.py)
+    python -m pip install -r requirements.txt
+    ```
+3. Build latest *StackClient* JAVA resource, copy `.jar` file and entire `lib` folder into `<tmp_stack>` repository, and install resource for py4jps (Please note that this requires [Java Runtime Environment version >=11]):
```
Comment (gpeb2) [inline Agents/MetOfficeAgent/README.md:181]:
```suggestion
3. Build latest *StackClient* JAVA resource, copy `.jar` file and entire `lib` folder into `<tmp_stack>` repository, and install resource for py4jps (Please note that this requires [Java Development Kit version >=11]):
```

File: Agents/MetOfficeAgent/README.md [hunk 4]
```diff
@@ -2,106 +2,117 @@
...
-The required steps are detailed in the [py4jps] documentation and already included in the agent's [Dockerfile]. <span style="color:red">Updating the py4jps resource is ONLY required if a pre-release version is needed. Otherwise, this step is irrelevant and the `build_py4jps_resources.sh` script shall be removed in the future</span>. Compiling those resources requires a [Java Runtime Environment version >=11].
+This agent requires [JPS_BASE_LIB] and [Stack-Clients] to be wrapped by [py4jps]. Therefore, after installation of all required packages (incl. `py4jps >= 1.0.26`), the `StackClients` resource needs to be added to allow for access through `py4jps`. All required steps are detailed in the [py4jps] documentation. However, the commands provided below shall suffice to compile the latest `StackClients` resource locally and install it inside the Docker container using the provided [Dockerfile]. Please note, that compiling requires a [Java Runtime Environment version >=11]. *Updating the [JPS_BASE_LIB] resource is ONLY required if a pre-release version is needed, which is (currently) not the case for this agent.*
```
Comment (gpeb2) [inline Agents/MetOfficeAgent/README.md:65]:
```suggestion
This agent requires [JPS_BASE_LIB] and [Stack-Clients] to be wrapped by [py4jps]. Therefore, after installation of all required packages (incl. `py4jps >= 1.0.26`), the `StackClients` resource needs to be added to allow for access through `py4jps`. All required steps are detailed in the [py4jps] documentation. However, the commands provided below shall suffice to compile the latest `StackClients` resource locally and install it inside the Docker container using the provided [Dockerfile]. Please note, that compiling requires a [Java Development Kit version >=11]. *Updating the [JPS_BASE_LIB] resource is ONLY required if a pre-release version is needed, which is (currently) not the case for this agent.*
```

File: Agents/MetOfficeAgent/README.md [hunk 5]
```diff
@@ -2,106 +2,117 @@
...
 
 ### **2) Accessing Github's Container registry**
 
-While building the Docker image of the agent, it also gets pushed to the [Container registry on Github]. Access needs to be ensured beforehand via your github [personal access token], which must have a `scope` that [allows you to publish and install packages]. To log in to the [Container registry on Github] simply run the following command to establish the connection and provide the access token when prompted:
+While building the Docker image of the agent, it also gets pushed to the [Github container registry]. Access needs to be ensured beforehand via your github [personal access token], which must have a `scope` that [allows you to publish and install packages]. To log in to the [Container registry on Github] simply run the following command to establish the connection and provide the access token when prompted:
 ```
-  $ docker login ghcr.io -u <github_username>
-  $ <github_personal_access_token>
+$ docker login ghcr.io -u <github_username>
+$ <github_personal_access_token>
```
Comment (gpeb2) [inline Agents/MetOfficeAgent/README.md:37]: Remove `$`s for consistency
```suggestion
docker login ghcr.io -u <github_username>
<github_personal_access_token>
```

File: Agents/MetOfficeAgent/README.md [hunk 6]
```diff
@@ -2,106 +2,117 @@
...
+Before starting development or spinning up the dockerized agent remotely, all required VSCode extensions shall be installed on the remote machine (e.g. *augustocdias.tasks-shell-input* or the *Python extension*). As the Docker image requires the [Stack-Clients] `.jar` file to be wrapped by [py4jps], it needa to be copied over manually to the respective folder as specified in the [Dockerfile] or can be created remotely by running the *Update JPSRM and Build and Debug* Debug Configuration. In order to build the resource, Java and Maven need to be available on the remote machine. In order to pull TWA specific Maven packages from the [Github package repository], `settings.xml` and `settings-security.xml` files need to be copied into Maven's `.m2` folder on the remote machine (typically located at user's root directory)
```
Comment (gpeb2) [inline Agents/MetOfficeAgent/README.md:105]:
```suggestion
Before starting development or spinning up the dockerized agent remotely, all required VSCode extensions shall be installed on the remote machine (e.g. *augustocdias.tasks-shell-input* or the *Python extension*). As the Docker image requires the [Stack-Clients] `.jar` file to be wrapped by [py4jps], it needs to be copied over manually to the respective folder as specified in the [Dockerfile] or can be created remotely by running the *Update JPSRM and Build and Debug* Debug Configuration. In order to build the resource, Java and Maven need to be available on the remote machine. In order to pull TWA specific Maven packages from the [Github package repository], `settings.xml` and `settings-security.xml` files need to be copied into Maven's `.m2` folder on the remote machine (typically located at user's root directory)
```

**Review by gpeb2 — CHANGES_REQUESTED** [review]
Issues in the README. Particularly addressing the confusion between the Java Runtime Environment (JRE), used to run Java code, and the Java Development Kit (JDK), used to compile Java code.

**Comment by markushofmeister** [comment]
> Issues in the README. Particularly addressing the confusion between the Java Runtime Environment (JRE), used to run Java code, and the Java Development Kit (JDK), used to compile Java code.

Should be addressed by [b0e72ea](https://github.com/cambridge-cares/TheWorldAvatar/pull/407/commits/b0e72ea05888d0961809cb1ebb17a8eb3e140630)

**Pull request merged**